### ISSUE TITLE
selector doesn't return all elements 

### ISSUE THREAD

**Issue by a-eid** [tmpvar/jsdom#1877]
### Basic info:

- **Node.js version:** 
` 8.0.0`
- **jsdom version:** 
`11.0.0`
### Minimal reproduction case

here is what chrome returns 

![](https://i.imgur.com/BTcg2Z2.png)

```js
const r = require('request')
const jsdom = require("jsdom");
const { JSDOM } = jsdom



r('http://www.clasificadosonline.com/Rentals.asp', function (err, response, body) {
  const dom = new JSDOM(body, {
    userAgent: "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36",
  });

  console.log(
    dom.window.document
      .querySelectorAll('tr[bgcolor="#ffffff"] a[href^="/PartnersListingREFRID.asp?"]')
      .length 
  )
})
```

here is what jsdom returns 

![](https://i.imgur.com/M9PCSyH.png) 

note that I've test requestjs response and it does include the complete HTML

**Comment by domenic** [comment]
Almost certainly what is happening here is that some elements are being added from script, and you are not running scripts.

**Comment by a-eid** [comment]
Yeah, that is what I thought, I have stripped every script tag in the requestjs response manually, opened it in Chrome and reperformed the query.

**Comment by domenic** [comment]
Hmm OK, thanks. Could you upload the stripped-of-script-tags version to a gist or pastebin or something? Any help moving it down to a small repro case of like a 20-line page would be even better. E.g. maybe you could analyze which elements are excluded and see if they have anything in common.

**Comment by a-eid** [comment]
here is the stripped gist.

**Comment by a-eid** [comment]
I think it be a`chrome's` problem not `jsodom's` 

now the selector is `'tr[bgcolor="#ffffff"] a[href^="/PartnersListingREFRID.asp?"]'` 

looking at the following two pics 

![](https://i.imgur.com/w5VZpND.png]
![](https://i.imgur.com/nv59Aqm.png] 

chrome detecting `1160` query result. 


now clearly looking at the HTML 

![](https://i.imgur.com/s7jz7Qt.png) 

no every `a[href^="/PartnersListingREFRID.asp?` is a descendent of `tr[bgcolor="#ffffff"] ` 

although it could be a parsing problem in either `chrome` or `jsdom` 

firefox results agrees with `chromes`

**Comment by Zirro** [comment]
From a quick test, it appears to have something to do with case-sensitivity. Replacing `#ffffff` with `#FFFFFF` in the selector yields 1160 nodes as with browsers. It is presumably related to [this logic](https://github.com/dperini/nwmatcher/blob/master/src/nwmatcher-noqsa.js#L544-#L556), though I don't have time to delve further into it today. Pinging @dperini, in case you'd like to have a look.

**Comment by a-eid** [comment]
Thanks, I always thought CSS attribute selectors are case-insensitive though .

**Comment by domenic** [comment]
Let's reopen, as being different from browsers is definitely a bug. They're not case-insensitive though unless you use the "i" modifier.

**Comment by Zirro** [comment]
The spec has a vague mention of different rules for case sensitivity based on [document language](https://drafts.csswg.org/selectors-4/#attribute-case) - potentially including quirks mode. That said, I see no difference in output from jsdom or browsers when adding a DOCTYPE.

**Comment by domenic** [comment]
Oh, I guess we or nwmatcher need to implement this: https://html.spec.whatwg.org/#case-sensitivity-of-selectors

**Comment by Zirro** [comment]
Seems like nwmatcher.js (but not nwmatcher-noqsa.js) already implements it: https://github.com/dperini/nwmatcher/blob/master/src/nwmatcher.js#L380. As I'm fairly unfamiliar with the project, it is not clear to me whether there's an intentional reason for that list being present in only one of the files.

**Comment by dperini** [comment]
@Zirro / @domenic you brought up a ancient problem, probably not easy to handle ...

Yes, the knowledge and distinction between Quirks Mode and Standard Mode is already embedded in "nwmatcher" but I am not sure browsers implement that correctly. I hope to be proved wrong and I am willing to work on that, also, I already tried hard in the past without success. 

At the time I implemented that table from: http://rakaz.nl/item/css_selector_bugs_case_sensitivity browsers were still buggy and didn't bother fixing that aspect of "Quirks Mode". Only with the advent of the HTML5 standards things started to change and browsers started to implement it.

I was basically forced to give up and bending my good will to the inconsistency and inaccuracy of browser vendors.

To make a long story short when I forked the "-noqsa" version the browsers were still buggy so to have the most accurate emulation I decided to implement that table partially in the "-noqsa" version by picking only the most popular attributes. See the INSENSITIVE_MAP: https://github.com/dperini/nwmatcher/blob/master/src/nwmatcher-noqsa.js#L122

I can port the complete table (46 attributes) to the "-noqsa" version but I then expect somebody else will be unhappy about the changed behavior and we will still have difference with some browser. So we will be back at "zero" problems solved again.

Please download the two test pages from this folder, I kept them ... waiting for the problem to resurrect and hit me back again:

    https://github.com/dperini/nwmatcher/tree/master/test/quirks/

now from the console execute the following queries, in each of the test pages, in Chrome and Firefox:

    document.querySelectorAll('#dom .foobar');
    document.querySelectorAll('#dom [class="foobar"]');
    document.querySelectorAll('#dom [class~="foobar"]');
    document.getElementById('dom').getElementsByClassName('foobar');

    document.querySelectorAll('#qsa .foobar');
    document.querySelectorAll('#qsa [class="foobar"]');
    document.querySelectorAll('#qsa [class~="foobar"]');
    document.getElementById('qsa').getElementsByClassName('foobar');

then try to draw your conclusions on this case sensitivity problem. Maybe you can find out were I am doing something wrong and can help me out fixing this longstanding bug once and forever.
Or ... could it be that "nwmatcher" implements Quirk/Standard Mode distinction correctly ?

Thank you for the report, for the patience to read and for digging in to this mud :9)

P.S. Slightly related, I just added Selectors Level 4 case sensitivity flag [class="MiXeD" i]

**Comment by Zirro** [comment]
@dperini That is interesting! Thanks for sharing the history around this issue.

From the spec, it looks like the list of attributes with case-insensitive values (which includes `bgcolor`) should apply regardless of standards/quirks mode. I created a testcase with the `bgcolor` attribute instead of `class` and this showed a consistent passing behaviour in Firefox, Chrome and Safari. The `class` attribute is not included in this list, however.

> `Or ... could it be that "nwmatcher" implements Quirk/Standard Mode distinction correctly ?`

The `class` attribute seems to be a special case. The spec says that "IDs and classes in no-quirks mode and limited-quirks mode" should be matched case-sensitively. I suppose that doesn't explicitly tell us that the match should be case-insensitive in quirks mode, but that has to be the intention given the context. That's also how nwmatcher works right now.

However, the browsers I tested in (Firefox, Chrome, Safari) seem to do a case-sensitive match on `class` attribute values even in quirks mode, contradicting the spec. I don't have IE/Edge available to test in right now, but if their behaviour matches the other browsers it would seem like the spec should be changed to say that `class` attribute values always should be matched case sensitively.

It is possible that confusion was derived from the following sentence, which is part of the same section:

> Selectors defines that ID and class selectors, when matched against elements in documents that are in quirks mode, will be matched in an ASCII case-insensitive manner.

I believe that refers only to `.foo` selectors, and not `[class="foo"]` selectors.

I'll check more thoroughly tomorrow whether I've got this right and file a spec bug if that turns out to be the case. I might also submit a PR to nwmatcher :-)

**Comment by dperini** [comment]
Excellent, you got to the right point:

> I believe that refers only to .foo selectors, and not [class="foo"] selectors.

The difference is that in Quirks Mode the `.foo` selector does not yeld the same results as the `[class="foo"]` selector. Also consider comparing the other alternative to collect elements with a specific class name: `document.getElementsByClassName("foo")`.

This is really confusing, anyway my hair were already gone a few years ago so that couldn't make things worst in that respect.

English is not my native language and even after reading the various specs a few times I couldn't find a clear statement saying the above but I am happy you could guess it from the sentence you mentioned above.

If that turns to be true I can easily import the full exception table even in "nwmatcher-noqsa" and mimic the browser way of always match the `class` attribute case sensitively. This will level up the current inconsistencies with the browsers in Quirks Mode.

Thank you for your time and precious help !

**Comment by dperini** [comment]
As a reference these are links to "nwmatcher" commits regarding the HTML_TABLE and older case sensitivity bugs and fixes. The HTML_TABLE is a list of 46 attributes names that need special handling due to case sensitivity bugs in various browsers. These links are ordered by descending dates (all from 2009):

https://github.com/dperini/nwmatcher/commit/060350ffd998c5894854a059b72ce185c1ee8b7b
https://github.com/dperini/nwmatcher/commit/47079733b571e6e1fac7bebde1de1f168622a4c9
https://github.com/dperini/nwmatcher/commit/0b222e8f142cc010c0d4cc1657c7236d78162ada
https://github.com/dperini/nwmatcher/commit/0b222e8f142cc010c0d4cc1657c7236d78162ada
https://github.com/dperini/nwmatcher/commit/e12acc50e2b85d4de65e0e2b4e247fb7f9c8a937
https://github.com/dperini/nwmatcher/commit/a77eb04bb5ca4b555f2efc5cdfed01ce1d064dcb

**Comment by dperini** [comment]
I have pushed the requested (case sensitivity) fix here: https://github.com/dperini/nwmatcher/commit/3a1cd75a by copying it from the complete "nwmatcher.js"  implementation.

Browsers do not agree with the specifications on how to match `class` attributes in Quirks Mode.

The HTML Living Standard (https://html.spec.whatwg.org/#case-sensitivity-of-selectors) is recommending a case insensitive match. This has been the same for the last 8/10 years now.

Since most browsers do not follow this specific rule, in Quirks Mode,  there will be inconsistencies with the results returned by "nwmatcher". Though due to extremely rare use of Quirks Mode the impact will probably be minimal.

I will wait for comments/suggestion and bugs review before pushing these changes to `npm`.

Please test and let me know if this solves the reported issue.

**Comment by Zirro** [comment]
Great work! However, I just [submitted a pull request](https://github.com/whatwg/html/pull/2737) to change the standard and clarify that `id` and `class` in attribute selectors should always match their values case-sensitively. The change ought to get accepted since that's what browsers actually do (thankfully, that's how the standards are written these days). You'll want to change nwmatcher to do that instead.

**Comment by dperini** [comment]
@Zirro I already applied the suggested change in this commit:
https://github.com/dperini/nwmatcher/commit/88b2278

Maybe I can take advantage of your knowledge and skills, when you see you have the time for it,
I would appreciate one last check to ensure everything is correct, so I can sleep over this case.

Thank you for helping out with this.

**Comment by dperini** [comment]
I should revert and re-apply the commit, it was wrong and not commented/described in the sources:
The latest commit is: https://github.com/dperini/nwmatcher/commit/0ab4041
I edited and corrected also my previous message.

**Comment by Zirro** [comment]
I'd be happy to help test the changes in detail, though it might take a day or two before I have time. I hope that's alright.

**Comment by dperini** [comment]
@Zirro sure that's perfect ... 
There is no real rush on this so take the time you need.
Thank you again for the help.
Have a nice day.

**Comment by Zirro** [comment]
Okay, the attributes are being treated correctly now when it comes to HTML documents. Great!

However, the specification mentions that the list only applies to HTML documents. When I've tested on XML/XHTML documents in Firefox, Chrome and Safari it seems like the attribute values are always treated case-sensitively. In other words, there should be no need for the XHTML_TABLE as they are always case-sensitive in XML.

I believe that's all that remains to match browsers now :-)

**Comment by dperini** [comment]
@Zirro thank you so much for the review !!!

I removed the XHTML_TABLE from the sources so for XML document the comparison will always be case sensitive also in "nwmatcher.js". Notice that in previous versions of "nwmatcher-noqsa.js", the one used in "jsdom", the behavior was already like that, the XHTML_TABLE has never been used.

@domenic I will wait a couple of days before pushing 1.4.1 to npm.

**Comment by domenic** [comment]
Looks like this is fixed! Master, which will be published shortly, has nwmatcher 1.4.1 as the new minimum version, so all new installs will get this.
